File: merge_csv_by_match.py
#!/usr/bin/env python3
"""Merge two CSV files by the `match` column.

For each row in the first CSV:
- if that row's `match` value exists in the second CSV, write the row from
  the second CSV to the output;
- otherwise, write the row from the first CSV to the output.
"""

from __future__ import annotations

import argparse
import csv
from pathlib import Path


def read_csv(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    with path.open(newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ValueError(f"{path} is missing a header row")
        rows = list(reader)
    return reader.fieldnames, rows


def build_second_index(
    rows: list[dict[str, str]], match_column: str, source: Path
) -> dict[str, dict[str, str]]:
    index: dict[str, dict[str, str]] = {}

    for row_number, row in enumerate(rows, start=2):
        if match_column not in row:
            raise ValueError(f"{source} does not contain a '{match_column}' column")

        match_value = row[match_column]
        if match_value in index:
            raise ValueError(
                f"{source} contains duplicate '{match_column}' values: {match_value!r} "
                f"(duplicate found on row {row_number})"
            )
        index[match_value] = row

    return index


def merge_rows(
    first_headers: list[str],
    first_rows: list[dict[str, str]],
    second_headers: list[str],
    second_rows: list[dict[str, str]],
    match_column: str,
    second_source: Path,
) -> tuple[list[str], list[dict[str, str]]]:
    if match_column not in first_headers:
        raise ValueError(f"first CSV does not contain a '{match_column}' column")
    if match_column not in second_headers:
        raise ValueError(f"second CSV does not contain a '{match_column}' column")

    second_index = build_second_index(second_rows, match_column, second_source)
    output_headers = first_headers + [
        header for header in second_headers if header not in first_headers
    ]

    output_rows: list[dict[str, str]] = []
    for row in first_rows:
        match_value = row[match_column]
        output_rows.append(second_index.get(match_value, row))

    return output_headers, output_rows


def write_csv(path: Path, headers: list[str], rows: list[dict[str, str]]) -> None:
    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=headers)
        writer.writeheader()
        for row in rows:
            writer.writerow({header: row.get(header, "") for header in headers})


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "For each row in the first CSV, write the matching row from the "
            "second CSV if present; otherwise write the first row."
        )
    )
    parser.add_argument("first_csv", type=Path, help="Path to the first CSV file")
    parser.add_argument("second_csv", type=Path, help="Path to the second CSV file")
    parser.add_argument("output_csv", type=Path, help="Path to the output CSV file")
    parser.add_argument(
        "--match-column",
        default="match",
        help="Column name used to match rows between files (default: match)",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    first_headers, first_rows = read_csv(args.first_csv)
    second_headers, second_rows = read_csv(args.second_csv)
    output_headers, output_rows = merge_rows(
        first_headers,
        first_rows,
        second_headers,
        second_rows,
        args.match_column,
        args.second_csv,
    )
    write_csv(args.output_csv, output_headers, output_rows)


if __name__ == "__main__":
    main()

File: merge_csv_by_match.php
<?php

declare(strict_types=1);

/**
 * Merge two CSV files by the "match" column.
 *
 * For each row in the first CSV:
 * - if that row's "match" value exists in the second CSV, write the row from
 *   the second CSV to the output;
 * - otherwise, write the row from the first CSV to the output.
 */

function readCsv(string $path): array
{
    $handle = fopen($path, 'rb');
    if ($handle === false) {
        throw new RuntimeException("Unable to open CSV file: {$path}");
    }

    $headers = fgetcsv($handle);
    if ($headers === false) {
        fclose($handle);
        throw new RuntimeException("{$path} is missing a header row");
    }

    $rows = [];
    while (($values = fgetcsv($handle)) !== false) {
        if ($values === [null] || $values === []) {
            continue;
        }

        $row = [];
        foreach ($headers as $index => $header) {
            $row[$header] = $values[$index] ?? '';
        }
        $rows[] = $row;
    }

    fclose($handle);

    return [$headers, $rows];
}

function buildSecondIndex(array $rows, string $matchColumn, string $sourcePath): array
{
    $index = [];

    foreach ($rows as $rowOffset => $row) {
        if (!array_key_exists($matchColumn, $row)) {
            throw new RuntimeException(
                "{$sourcePath} does not contain a '{$matchColumn}' column"
            );
        }

        $matchValue = $row[$matchColumn];
        if (array_key_exists($matchValue, $index)) {
            $rowNumber = $rowOffset + 2;
            throw new RuntimeException(
                "{$sourcePath} contains duplicate '{$matchColumn}' values: '{$matchValue}' " .
                "(duplicate found on row {$rowNumber})"
            );
        }

        $index[$matchValue] = $row;
    }

    return $index;
}

function mergeRows(
    array $firstHeaders,
    array $firstRows,
    array $secondHeaders,
    array $secondRows,
    string $matchColumn,
    string $secondSourcePath
): array {
    if (!in_array($matchColumn, $firstHeaders, true)) {
        throw new RuntimeException("first CSV does not contain a '{$matchColumn}' column");
    }
    if (!in_array($matchColumn, $secondHeaders, true)) {
        throw new RuntimeException("second CSV does not contain a '{$matchColumn}' column");
    }

    $secondIndex = buildSecondIndex($secondRows, $matchColumn, $secondSourcePath);
    $outputHeaders = $firstHeaders;
    foreach ($secondHeaders as $header) {
        if (!in_array($header, $outputHeaders, true)) {
            $outputHeaders[] = $header;
        }
    }

    $outputRows = [];
    foreach ($firstRows as $row) {
        $matchValue = $row[$matchColumn];
        $outputRows[] = $secondIndex[$matchValue] ?? $row;
    }

    return [$outputHeaders, $outputRows];
}

function writeCsv(string $path, array $headers, array $rows): void
{
    $handle = fopen($path, 'wb');
    if ($handle === false) {
        throw new RuntimeException("Unable to write CSV file: {$path}");
    }

    fputcsv($handle, $headers);
    foreach ($rows as $row) {
        $orderedRow = [];
        foreach ($headers as $header) {
            $orderedRow[] = $row[$header] ?? '';
        }
        fputcsv($handle, $orderedRow);
    }

    fclose($handle);
}

function printUsageAndExit(): never
{
    fwrite(
        STDERR,
        "Usage: php merge_csv_by_match.php first.csv second.csv output.csv [match-column]\n"
    );
    exit(1);
}

function main(array $argv): void
{
    if (count($argv) < 4 || count($argv) > 5) {
        printUsageAndExit();
    }

    $firstCsv = $argv[1];
    $secondCsv = $argv[2];
    $outputCsv = $argv[3];
    $matchColumn = $argv[4] ?? 'match';

    [$firstHeaders, $firstRows] = readCsv($firstCsv);
    [$secondHeaders, $secondRows] = readCsv($secondCsv);
    [$outputHeaders, $outputRows] = mergeRows(
        $firstHeaders,
        $firstRows,
        $secondHeaders,
        $secondRows,
        $matchColumn,
        $secondCsv
    );

    writeCsv($outputCsv, $outputHeaders, $outputRows);
}

if (PHP_SAPI === 'cli' && isset($_SERVER['SCRIPT_FILENAME']) && realpath($_SERVER['SCRIPT_FILENAME']) === __FILE__) {
    try {
        main($argv);
    } catch (Throwable $error) {
        fwrite(STDERR, $error->getMessage() . PHP_EOL);
        exit(1);
    }
}
